optaplanner-core/src/main/java/org/optaplanner/core/impl/score/buildin/bendable/BendableScoreInliner.java

```diff
@@ -55,31 +53,27 @@ public IntWeightedScoreImpacter buildWeightedScoreImpacter(BendableScore constra
             int levelWeight = constraintWeight.getHardOrSoftScore(singleLevel);
             if (singleLevel < constraintWeight.getHardLevelsSize()) {
                 int level = singleLevel;
-                return (int matchWeight, Consumer<Score<?>> matchScoreConsumer) -> {
+                return (int matchWeight, Supplier<List<Object>> justifications) -> {
```

Comment:
Justification list is a `Supplier` - this way, the collection will only ever be initialized (with the perf impact) when we need justifications. Therefore, when constraint match is disabled, justifications have literally zero impact.